### PR TITLE
COMPAT: Adapt to Numpy 2.0 dtype changes

### PR DESCRIPTION
**Pull request by seberg**
Based on Thomas Li's work, but wanted to try and see that this is right.

Replaces gh-13466.

### CODE REVIEW

File: pandas/_libs/src/datetime/pd_datetime.c [hunk 1]
```diff
@@ -255,5 +257,6 @@ static struct PyModuleDef pandas_datetimemodule = {
 
 PyMODINIT_FUNC PyInit_pandas_datetime(void) {
   PyDateTime_IMPORT;
+  import_array();
```
Comment (WillAyd) [inline pandas/_libs/src/datetime/pd_datetime.c:260]: So this needs to be called any time you import `ndarrayobject.h` right? If so, I wonder if it wouldn't be cleaner to move the include here to the header. That way other module that need these datetime functions can simply include / link the pd_datetime capsule and not have to include ndarrayobject.h directly. May be less confusing

Comment (seberg) [reply pandas/_libs/src/datetime/pd_datetime.c:261]: Yes (if you use any API, but if you don't you only need `ndarraytypes.h`) but once per compilation unit only (just fixed that).
Places that use `pd_datetime.h` don't need it (we just need to be sure the init code ran).

**Review by lithomas1 — APPROVED** [review]

**Review by WillAyd — APPROVED** [review]

**Comment by WillAyd** [comment]
@lithomas1 this needs backport right?

**Comment by lithomas1** [comment]
Mhm

**Pull request merged**